tools/archiver.py

```diff
@@ -482,15 +482,31 @@ def default_empty_string(value):
             message_date = email.utils.parsedate_tz(
                 str(msg_metadata.get("archived-at"))
             )
-
         if not message_date:
-            epoch = time.time()
+            print("No message date could be derived from the Date: header, looking elsewhere.")
+            # See if we have a "From" header line in the raw email, we can use
+            first_line = raw_msg.split(b"\n", 1)[0].decode("us-ascii")
+            if first_line.startswith("From "):
+                # If we have one, the date must be the third element when splitting by single space.
+                env_from_date = first_line.split(" ", 2)[-1]  # Split twice, grab last element.
+                message_date = email.utils.parsedate_tz(env_from_date)
+                if message_date:
+                    print("Found date in envelope FROM header: %s" % env_from_date)
+            # Otherwise, look for a Received: header we can scan
+            if not message_date:
+                for recv_from in msg.get_all('received', []):  # We may have multiple of these, not all have "from".
+                    m = re.match(r"from[^;]+?;\s+(.+?)(?:$|[\r\n])", recv_from)
+                    if m:
+                        message_date = email.utils.parsedate_tz(m.group(1))
+                        if message_date:
+                            print("Found date in Received header: %s" % m.group(1))
+                            break
+            if not message_date:
+                print("Could not find any valid dates in email headers, using current time")
+                epoch = time.time()
```

Comment:
Is the current time the correct fallback here?
If so, then add an explanatory comment to the code.

Reply:
A time of some sort must be used. The present time makes most sense as a default, as that would work for archiving in real-time. For importing, there really isn't a proper value either way. Both epoch 0 and current would be equally "wrong".

Reply:
For archiving, I agree that current time is OK.
For importing, I think it would be better to use a value that does not vary between imports.

Reply:
For importing maybe there should be a option to skip the message; it may make sense to allow the user to correct the data - especially since they can add a suitable From time without needing to change the message content.